<h1 align = center>Introduction to Neural Networks and Deep Learning with Keras <br>
    <br>
    Final Project </h1>


<h2> Data Preparation<h2>

In [1]:
#Import libraries to prep data
import numpy as np
import pandas as pd

In [2]:
#Load dataset
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
#Display dimension
concrete_data.shape

(1030, 9)

In [4]:
#Check for missing values
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [5]:
#Split data into features and target values
concrete_columns = concrete_data.columns

features = concrete_data[concrete_columns[concrete_columns != 'Strength']]
target = concrete_data['Strength']

In [6]:
features.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [7]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

<h2>A. Build a baseline model</h2>


In [8]:
#Import Keras
import keras
from keras.models import Sequential
from keras.layers import Dense

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

Using TensorFlow backend.


In [9]:
#Build Neural Network for Regression
def regression():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(features.shape[1],)))
    model.add(Dense(1))
    
    #Define optimizer and loss function
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    
    return model
    

In [10]:
#Create model
reg_model = regression()

Instructions for updating:
Colocations handled automatically by placer.


In [13]:
mse_list = []

#50 iterations
for i in range(50):
    
    #Train Test Split
    feat_train, feat_test, target_train, target_test = train_test_split( features, target, test_size=0.3, random_state=4)
    
    #Train model
    reg_model.fit(feat_train, target_train, epochs=50)
    
    #Prediction on test data
    target_pred = reg_model.predict(feat_test)
    
    #Compute MSE for evaluation by comparing prediction and actual value
    mse = mean_squared_error(target_test, target_pred)
    mse_list.append(mse)
    print("Mean Squared Error:", mse)    

Instructions for updating:
Use tf.cast instead.
Epoch 1/50
721/721 [==============================] - 1s 1ms/step - loss: 4078.4950
Epoch 2/50
721/721 [==============================] - 0s 244us/step - loss: 3064.7767
Epoch 3/50
721/721 [==============================] - 0s 274us/step - loss: 2271.3651
Epoch 4/50
721/721 [==============================] - 0s 246us/step - loss: 1722.1885
Epoch 5/50
721/721 [==============================] - 0s 254us/step - loss: 1407.5408
Epoch 6/50
721/721 [==============================] - 0s 224us/step - loss: 1205.4844
Epoch 7/50
721/721 [==============================] - 0s 664us/step - loss: 1051.3464
Epoch 8/50
721/721 [==============================] - 0s 256us/step - loss: 920.3449
Epoch 9/50
721/721 [==============================] - 0s 248us/step - loss: 818.7888
Epoch 10/50
721/721 [==============================] - 0s 253us/step - loss: 737.1152
Epoch 11/50
721/721 [==============================] - 0s 236us/step - loss: 661.5768
Epoch 12/5

In [15]:
#Calculate Mean and Satndard Deviation
mse_mean = np.mean(mse_list)
print("Mean of the MSE: ", mse_mean )

std = np.std(mse_list)
print("Standard Deviation of the MSE: ", std)

Mean of the MSE:  56.999975342284195
Standard Deviation of the MSE:  21.093072070768702


<h2>B. Normalize the data<h2>

In [17]:
#Normalize data
feat_norm = (features - features.mean()) / features.std()
feat_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [18]:
mse_list = []

#50 iterations
for i in range(50):
    
    #Train Test Split
    feat_train, feat_test, target_train, target_test = train_test_split( feat_norm, target, test_size=0.3, random_state=4)
    
    #Train model
    reg_model.fit(feat_train, target_train, epochs=50)
    
    #Prediction on test data
    target_pred = reg_model.predict(feat_test)
    
    #Compute MSE for evaluation by comparing prediction and actual value
    mse = mean_squared_error(target_test, target_pred)
    mse_list.append(mse)
    print("Mean Squared Error:", mse)    

Epoch 1/50
721/721 [==============================] - 0s 228us/step - loss: 1561.8233
Epoch 2/50
721/721 [==============================] - 0s 234us/step - loss: 1480.4708
Epoch 3/50
721/721 [==============================] - 0s 240us/step - loss: 1399.9556
Epoch 4/50
721/721 [==============================] - 0s 247us/step - loss: 1321.2388
Epoch 5/50
721/721 [==============================] - 0s 246us/step - loss: 1242.1896
Epoch 6/50
721/721 [==============================] - 0s 230us/step - loss: 1165.7743
Epoch 7/50
721/721 [==============================] - 0s 276us/step - loss: 1092.2118
Epoch 8/50
721/721 [==============================] - 0s 218us/step - loss: 1021.2825
Epoch 9/50
721/721 [==============================] - 0s 239us/step - loss: 953.3207
Epoch 10/50
721/721 [==============================] - 0s 227us/step - loss: 887.9578
Epoch 11/50
721/721 [==============================] - 0s 254us/step - loss: 825.8132
Epoch 12/50
721/721 [==============================] - 

In [19]:
#Calculate Mean and Satndard Deviation
mse_mean = np.mean(mse_list)
print("Mean of the MSE: ", mse_mean )

std = np.std(mse_list)
print("Standard Deviation of the MSE: ", std)

Mean of the MSE:  52.00467639929596
Standard Deviation of the MSE:  30.262179807987287


Compared to A the mean of the MSEs decresed marginally (from 56.999975342284195 to 52.00467639929596), whereas there was a significant increse of the dtandard deviation (from 21.093072070768702 to 30.262179807987287).

<h1>C. Increate the number of epochs</h1>

In [20]:
mse_list = []

#50 iterations
for i in range(50):
    
    #Train Test Split
    feat_train, feat_test, target_train, target_test = train_test_split( feat_norm, target, test_size=0.3, random_state=4)
    
    #Train model
    reg_model.fit(feat_train, target_train, epochs=100)
    
    #Prediction on test data
    target_pred = reg_model.predict(feat_test)
    
    #Compute MSE for evaluation by comparing prediction and actual value
    mse = mean_squared_error(target_test, target_pred)
    mse_list.append(mse)
    print("Mean Squared Error:", mse)    

Epoch 1/100
721/721 [==============================] - 1s 891us/step - loss: 29.2206
Epoch 2/100
721/721 [==============================] - 0s 250us/step - loss: 29.2073
Epoch 3/100
721/721 [==============================] - 0s 209us/step - loss: 29.2178
Epoch 4/100
721/721 [==============================] - 1s 719us/step - loss: 29.1901
Epoch 5/100
721/721 [==============================] - 0s 274us/step - loss: 29.2112
Epoch 6/100
721/721 [==============================] - 0s 298us/step - loss: 29.1879
Epoch 7/100
721/721 [==============================] - 0s 204us/step - loss: 29.2468
Epoch 8/100
721/721 [==============================] - 0s 275us/step - loss: 29.1951
Epoch 9/100
721/721 [==============================] - 0s 213us/step - loss: 29.2803
Epoch 10/100
721/721 [==============================] - 0s 221us/step - loss: 29.2449
Epoch 11/100
721/721 [==============================] - 0s 220us/step - loss: 29.2357
Epoch 12/100
721/721 [==============================] - 0s 319u

In [21]:
#Calculate Mean and Satndard Deviation
mse_mean = np.mean(mse_list)
print("Mean of the MSE: ", mse_mean )

std = np.std(mse_list)
print("Standard Deviation of the MSE: ", std)

Mean of the MSE:  39.787764634032555
Standard Deviation of the MSE:  0.14422423824419373


Compared to B, both the mean of the MSE and standard deviation decresed significantly (Mean: from 52.00467639929596 to 39.787764634032555, Std: from 30.262179807987287 to 0.14422423824419373).

<h1>D. Increase the number of hidden layers</h1>

In [22]:
#Build Neural Network for Regression
def regression2():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(features.shape[1],)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    #Define optimizer and loss function
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    
    return model
    

In [23]:
#Create model
reg_model2 = regression2()

In [24]:
mse_list = []

#50 iterations
for i in range(50):
    
    #Train Test Split
    feat_train, feat_test, target_train, target_test = train_test_split( feat_norm, target, test_size=0.3, random_state=4)
    
    #Train model
    reg_model2.fit(feat_train, target_train, epochs=50)
    
    #Prediction on test data
    target_pred = reg_model2.predict(feat_test)
    
    #Compute MSE for evaluation by comparing prediction and actual value
    mse = mean_squared_error(target_test, target_pred)
    mse_list.append(mse)
    print("Mean Squared Error:", mse)    

Epoch 1/50
721/721 [==============================] - 2s 3ms/step - loss: 1558.4922
Epoch 2/50
721/721 [==============================] - 0s 555us/step - loss: 1531.5906
Epoch 3/50
721/721 [==============================] - 0s 553us/step - loss: 1489.6528
Epoch 4/50
721/721 [==============================] - 0s 578us/step - loss: 1421.8682
Epoch 5/50
721/721 [==============================] - 0s 524us/step - loss: 1319.5502
Epoch 6/50
721/721 [==============================] - 0s 528us/step - loss: 1162.2553
Epoch 7/50
721/721 [==============================] - 0s 556us/step - loss: 931.4082
Epoch 8/50
721/721 [==============================] - 0s 555us/step - loss: 644.0542
Epoch 9/50
721/721 [==============================] - 0s 582us/step - loss: 402.9726
Epoch 10/50
721/721 [==============================] - 1s 1ms/step - loss: 274.6504
Epoch 11/50
721/721 [==============================] - 1s 1ms/step - loss: 229.2274
Epoch 12/50
721/721 [==============================] - 0s 566us

In [25]:
#Calculate Mean and Satndard Deviation
mse_mean = np.mean(mse_list)
print("Mean of the MSE: ", mse_mean )

std = np.std(mse_list)
print("Standard Deviation of the MSE: ", std)

Mean of the MSE:  44.272228810843444
Standard Deviation of the MSE:  20.362655503694427


Compared to B the mean of the MSEs decresed (from 52.00467639929596 to 44.272228810843444) as well as the standard deviation (from 30.262179807987287 to 20.362655503694427).